lib/schema.js

```diff
@@ -1345,45 +1345,99 @@ Schema.prototype.indexes = function() {
   var indexes = [];
   var schemaStack = [];
 
-  var collectIndexes = function(schema, prefix) {
+  /*!
+   * Checks for indexes added to subdocs using Schema.index().
+   * These indexes need their paths prefixed properly.
+   *
+   * schema._indexes = [ [indexObj, options], [indexObj, options] ..]
+   */
+  function fixSubIndexPaths(schema, prefix) {
+    var subindexes = schema._indexes,
+        len = subindexes.length,
+        indexObj,
+        newindex,
+        klen,
+        keys,
+        key,
+        subIndex = 0,
+        keyIndex;
+
+    for (subIndex = 0; subIndex < len; ++subIndex) {
+      indexObj = subindexes[subIndex][0];
+      keys = Object.keys(indexObj);
+      klen = keys.length;
+      newindex = {};
+
+      // use forward iteration, order matters
+      for (keyIndex = 0; keyIndex < klen; ++keyIndex) {
+        key = keys[keyIndex];
+        newindex[prefix + key] = indexObj[key];
+      }
+
+      indexes.push([newindex, subindexes[subIndex][1]]);
+    }
+  }
+
+  function collectIndexes(schema, prefix, optional) {
+    var keyIndex,
+        isObject,
+        options,
+        field,
+        index,
+        type,
+        path,
+        keys,
+        key;
+
     // Ignore infinitely nested schemas, if we've already seen this schema
     // along this path there must be a cycle
     if (schemaStack.indexOf(schema) !== -1) {
       return;
     }
+
     schemaStack.push(schema);
 
     prefix = prefix || '';
-    var key, path, index, field, isObject, options, type;
-    var keys = Object.keys(schema.paths);
+    keys = Object.keys(schema.paths);
 
-    for (var i = 0; i < keys.length; ++i) {
-      key = keys[i];
+    for (keyIndex = 0; keyIndex < keys.length; ++keyIndex) {
+      key = keys[keyIndex];
       path = schema.paths[key];
 
-      if ((path instanceof MongooseTypes.DocumentArray) || path.$isSingleNested) {
-        collectIndexes(path.schema, prefix + key + '.');
+      if (path instanceof MongooseTypes.DocumentArray || path.$isSingleNested) {
+        collectIndexes(path.schema, prefix + key + '.', !path.isRequired || optional);
```

Comment:
Problem: `isRequired` is not necessarily known until you create a document, `required: function() { return false; }` is also supported.